corehq/util/pagination.py

```diff
@@ -0,0 +1,272 @@
+        """
+        raise StopIteration
+
+
+def paginate_function(data_function, args_provider, event_handler=PaginationEventHandler()):
+    """
+    Repeatedly call a data provider function with successive sets of arguments provided
+    by the ``args_provider``
+
+    :param data_function: function to paginate. Must return an list of data elements.
+    :param args_provider: An instance of the ``ArgsProvider`` class.
+    :param event_handler: class to be notified on page start and page end.
+    """
+    total_emitted = 0
+    args, kwargs = args_provider.get_initial_args()
+    while True:
+        event_handler.page_start(total_emitted, *args, **kwargs)
+        results = data_function(*args, **kwargs)
+        start_time = datetime.utcnow()
+        len_results = len(results)
+
+        for item in results:
+            yield item
+
+        total_emitted += len_results
+        event_handler.page_end(total_emitted, datetime.utcnow() - start_time, *args, **kwargs)
+
+        item = item if len_results else None
+        try:
+            args, kwargs = args_provider.get_next_args(item, *args, **kwargs)
+        except StopIteration:
+            break
+
+
+class ResumableIteratorState(JsonObject):
+    doc_type = "ResumableIteratorState"
+    _id = StringProperty()
+    name = StringProperty()
+    timestamp = DateTimeProperty()
+    args = ListProperty()
+    kwargs = DictProperty()
+    retry = DictProperty()
+    progress = DictProperty()
+    complete = BooleanProperty(default=False)
+
+
+class ResumableArgsProvider(ArgsProvider):
+    def __init__(self, iterator_state, args_provider):
+        self.args_provider = args_provider
+        self.resume = bool(getattr(iterator_state, '_rev', None))  # if there is a _rev then we're resuming
+        self.resume_args = iterator_state.args
+        self.resume_kwargs = iterator_state.kwargs
+
+    def get_initial_args(self):
+        if self.resume:
+            return self.resume_args, self.resume_kwargs
+        return self.args_provider.get_initial_args()
+
+    def get_next_args(self, last_item, *last_args, **last_kwargs):
+        return self.args_provider.get_next_args(last_item, *last_args, **last_kwargs)
+
+
+class ResumableFunctionIterator(object):
+    """Perform one-time resumable iteration over a data provider function.
+
+    Iteration can be efficiently stopped and resumed.
+
+    :param iteration_key: A unique key identifying the iteration. This
+    key will be used in combination with `iteration_function` name to maintain state
+    about an iteration that is in progress. The state will be maintained
+    indefinitely unless it is removed with `discard_state()`.
+    :param data_function: function to iterate over. Must return an list of data elements.
+    :param args_provider: An instance of the ``ArgsProvider`` class.
+    :param item_getter: Function which can be used to get an item by ID. Used for retrying items that failed.
```

Comment:
This complicates the constructor by adding a required argument that is only used by the optional retry feature. The generalization in this PR is making it less obvious how to use the retry feature: it requires this new `item_getter` and the doc ID must now be passed to `retry` instead of the object yielded from the iterator (doc).

It occurs to me that the retry feature could be implemented as a separate iterator/wrapper rather than conflating it with the resumable iterator. Usage would be something like this:
```python
def save_retries(data):
    # NOTE any other use of `resumable.progress_info` needs to
    # be aware of this so it doesn't stomp on the retry state.
    # To make that less error prone, could change progress_info
    # to be functions:
    #   resumable.get_progress_info(key)
    #   resumable.set_progress_info(key, value)
    progress_info = resumable.progress_info or {}
    progress_info["retry_data"] = data
    resumable.progress_info = progress_info

def item_getter(doc_id):
    try:
        return db.get(doc_id)
    except ResourceNotFound:
        raise SkipRetry

resumable = ResumableFunctionIterator(key, data_function, args_provider)
retries = (resumable.progress_info or {}).get("retries")

# `retries` and `save_retries` are optional arguments that can be
# omitted if persistence is not needed
retry_iter = RetryIterator(resumable, item_getter, retries, save_retries)

for doc in retry_iter:
    try:
        process(doc)
    except StaleDocError:
        retry_iter.retry(doc["_id"])
```